### PR TITLE
Clean up type checking related to renderMessages.js.

### PR DESCRIPTION
**Pull request by jackrzhang**
My initial goal was to add a type definition to annotate the `renderedMessages` prop passed from `MessageList` into `MessageListWeb`, which caused a number of flow error messages leading to this series of changes.

The data flow around this area is buggy - see #2627, #2676, and #2677 - so hopefully this elucidates the current structure of the subsystem and provides insight into any possibly unintended behavior.

### CODE REVIEW

File: src/types.js [hunk 1]
```diff
@@ -489,18 +489,6 @@ export type GetState = () => GlobalState;
 
 export type LocalizableText = any; // string | { text: string, values: Object };
 
-export type RenderedTimeDescriptor = {
```
Comment (borisyankov) [inline src/types.js]: I am not sure you need to delete these items in a separate commit (even though they are indeed not used)
In the currently last commit, you are reintroducing them under slightly different names.

Just use the last commit to replace these and skip this commit altogether.

File: src/types.js [hunk 2]
```diff
@@ -111,6 +111,7 @@ export type Message = {
   id: number,
   isOutbox: boolean,
   is_me_message: boolean,
+  last_edit_timestamp: number,
```
Comment (borisyankov) [inline src/types.js]: This prop is optional.

File: src/types.js [hunk 3]
```diff
@@ -54,6 +54,11 @@ export type Account = Auth;
  *
  * See also EventReaction, which carries additional properties computed from
  * these.
+ *
+ * To compute emoji_code from emoji_name, use the mapping exported
+ * by `src/emoji/codePointMap`.
+ *
+ * TODO: Add note on how to determine `reaction_type`.
```
Comment (borisyankov) [inline src/types.js]: Maybe not leave the TODO in the commit, but do describe the `reaction_type` computation.

File: src/message/__tests__/renderMessages-test.js [hunk 4]
```diff
@@ -2,13 +2,14 @@ import deepFreeze from 'deep-freeze';
 
 import { homeNarrow } from '../../utils/narrow';
 import renderMessages from '../renderMessages';
+import { NULL_MESSAGE } from '../../nullObjects';
```
Comment (borisyankov) [inline src/message/__tests__/renderMessages-test.js]: Great! I like this commit a lot :)

File: src/types.js (hunk 2 above)
Comment (borisyankov) [reply src/types.js]: If the `EditMessage` type does not make sense, just make the prop optional with `?`.

Comment (jackrzhang) [reply src/types.js]: Hm, okay. I'll look into the code paths for editing messages.

**Comment by gnprice** [comment]
Thanks for these cleanups!

One thing I'd like us to do, while we're here studying these and figuring out what these types really end up being: I'd like to record a way that the next person to look at the code can look up for themselves what these are supposed to be, and also can go check if there've been changes on the server.

Ideally this would be API documentation for Zulip. But there isn't yet much in the way of that, and not for the stuff like `Message` where we really most need it. So the jsdoc on some of these types is going to be a little more complicated than that.

I think we still want it, for a lot of these types, to look like: this represents the data from such-and-such events the server sends. Then instead of docs, we'll just have to say where you look in the server code to find that.

(This thought comes especially to mind on the first commit in this series, with the `last_edit_timestamp` addition and the `timeEdited` change.)

**Comment by gnprice** [comment]
BTW I spent a bit of time on this earlier today; and I've just pushed 563c9ae4e, which adds a big block comment on `Message` and some notes on specifics.

Still more to say on many more specifics! Like that `last_edit_timestamp`. Given the framework suggested in that block comment, on a commit like that I'd like to see a word about how you identified that this was the right type.

**Comment by gnprice** [comment]
Thanks again @jackrzhang for this PR!

At this point it looks like the improvements this made have all gotten merged in one way or another, so closing.

**Pull request closed without merging**